File: challenges/array/min_steps_infinite_grid.py


def cover_points(points):
    """
    Find the minimum number of steps to cover a sequence of points in the
    order they need to be covered.
    The points are in an infinite 2D grid and one can move in any of the
    8 directions.

        points = [(0,0, (2,2), (0,5)]

          0 1 2
        0 *
        1   \
        2     *
        3   /
        4 |
        5 *
    """
    moves = 0
    for prev_point, point in zip(points, points[1:]):
        prev_x, prev_y = prev_point
        x, y = point

        x_dist = abs(x - prev_x)
        y_dist = abs(y - prev_y)
        moves += max(x_dist, y_dist)

    return moves


def test_min_steps_cover_points():
    assert cover_points([(0, 0), (-2, 2)]) == 2
    assert cover_points([(0, 0), (-4, 2), (0, -2)]) == 8
    assert cover_points([(0, 0), (1, 1), (5, 2)]) == 5

File: challenges/array/min_steps_infinite_grid.cpp

#include <vector>
#include <cassert>
#include <algorithm>
#include <iostream>
#include <unordered_map>

typedef std::tuple<int, int> t_point;

int dist(t_point p1, t_point p2) {
    const auto prev_x = std::get<0>(p1);
    const auto prev_y = std::get<1>(p1);

    const auto cur_x = std::get<0>(p2);
    const auto cur_y = std::get<1>(p2);

    const auto dist_x = std::abs(cur_x - prev_x);
    const auto dist_y = std::abs(cur_y - prev_y);

    return std::max(dist_x, dist_y);
}

int cover_points(const std::vector<t_point> points) {
    int min_steps {0};

    for (auto it=points.begin(); it!=points.end(); ++it) {
        if(it == points.begin()) continue;

        min_steps += dist(*std::prev(it, 1), *it);
    }
    return min_steps;
}

int main() {
    std::vector<t_point> e1 = {
         std::make_tuple(0, 0),
         std::make_tuple(-2, 2)
    };
    assert(cover_points(e1) == 2);

    std::vector<t_point> e2 = {
         std::make_tuple(0, 0),
         std::make_tuple(-4, 2),
         std::make_tuple(0, -2)
    };
    assert(cover_points(e2) == 8);

    std::vector<t_point> e3 = {
         std::make_tuple(0, 0),
         std::make_tuple(1, 1),
         std::make_tuple(5, 2)
    };
    assert(cover_points(e3) == 5);
}
